merkle/src/merkle.rs

```diff
@@ -248,6 +258,241 @@ impl<E: Element> Clone for MmapStore<E> {
+        unimplemented!()
+    }
+}
+
+impl<E: Element> Store<E> for DiskMmapStore<E> {
+    #[allow(unsafe_code)]
+    fn new(size: usize) -> Self {
+        let byte_len = E::byte_len() * size;
+        let file: File = tempfile().expect("couldn't create temp file");
+        file.set_len(byte_len as u64)
+            .expect(&format!("couldn't set len of {}", byte_len));
+
+        let mmap = unsafe { MmapMut::map_mut(&file).expect("couldn't create map_mut") };
+        let mmap_size = mmap.len();
+        DiskMmapStore {
+            store: Arc::new(RwLock::new(Some(mmap))),
+            len: 0,
+            _e: Default::default(),
+            file,
+            store_size: mmap_size,
+            path: "".to_string(),
+            size: None,
+        }
+    }
+
+    fn new_from_slice(size: usize, data: &[u8]) -> Self {
+        assert_eq!(data.len() % E::byte_len(), 0);
+
+        let mut res = Self::new(size);
+
+        let end = data.len();
+        res.store_copy_from_slice(0, end, data);
+        res.len = data.len() / E::byte_len();
+
+        res
+    }
+
+    fn write_at(&mut self, el: E, i: usize) {
+        let b = E::byte_len();
+        self.store_copy_from_slice(i * b, (i + 1) * b, el.as_ref());
+        self.len += 1;
+    }
+
+    fn read_at(&self, i: usize) -> E {
+        let b = E::byte_len();
+        let start = i * b;
+        let end = (i + 1) * b;
+        let len = self.len * b;
+        assert!(start < len, "start out of range {} >= {}", start, len);
+        assert!(end <= len, "end out of range {} > {}", end, len);
+
+        E::from_slice(&self.store_read_range(start, end))
+    }
+
+    fn read_range(&self, r: ops::Range<usize>) -> Vec<E> {
+        let b = E::byte_len();
+        let start = r.start * b;
+        let end = r.end * b;
+        let len = self.len * b;
+        assert!(start < len, "start out of range {} >= {}", start, len);
+        assert!(end <= len, "end out of range {} > {}", end, len);
+
+        self.store_read_range(start, end)
+            .chunks(b)
+            .map(E::from_slice)
+            .collect()
+    }
+
+    fn len(&self) -> usize {
+        self.len
+    }
+
+    fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    fn push(&mut self, el: E) {
+        let l = self.len;
+        assert!(
+            (l + 1) * E::byte_len() <= self.store_size(),
+            format!(
+                "not enough space, l: {}, E size {}, store len {}",
+                l,
+                E::byte_len(),
+                self.store_size()
+            )
+        );
+
+        self.write_at(el, l);
+    }
+
+    // Offload the `store` in the case it was constructed with `new_with_path`.
+    // Temporary files with no path (created from `new`) can't be offloaded.
+    fn try_offload(&self) -> bool {
+        if self.path.is_empty() {
+            // Temporary file.
+            return false;
+        }
+
+        *self.store.write().unwrap() = None;
```

Comment:
Shouldn't this call `flush` on the mmap store?

Reply:
I _think_ that `msync` is implicit in `munmap`, at least for our [`MAP_SHARED`](https://linux.die.net/man/2/munmap) case.

Reply:
hmm looking at the example: https://docs.rs/memmap/0.7.0/memmap/struct.MmapMut.html#method.flush I am not sure, maybe we could call it just to make sure?